exporter/awsemfexporter/cwlog_client.go

```diff
@@ -0,0 +1,237 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package awsemfexporter
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/cloudwatchlogs"
+	"github.com/aws/aws-sdk-go/service/cloudwatchlogs/cloudwatchlogsiface"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsemfexporter/handler"
+)
+
+const (
+	// this is the retry count, the total attempts would be retry count + 1 at most.
+	defaultRetryCount          = 1
+	ErrCodeThrottlingException = "ThrottlingException"
+)
+
+var (
+	// backoff retry 2 times
+	sleeps = []time.Duration{
+		time.Millisecond * 200, time.Millisecond * 400}
+)
+
+//The log client will perform the necessary operations for publishing log events use case.
+type LogClient interface {
+	PutLogEvents(input *cloudwatchlogs.PutLogEventsInput, retryCnt int) (*string, error)
+	CreateStream(logGroup, streamName *string) (token string, e error)
+}
+
+// Possible exceptions are combination of common errors (https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/CommonErrors.html)
+// and API specific erros (e.g. https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_PutLogEvents.html#API_PutLogEvents_Errors)
+type cloudWatchLogClient struct {
+	svc    cloudwatchlogsiface.CloudWatchLogsAPI
+	logger *zap.Logger
+}
+
+//Create a log client based on the actual cloudwatch logs client.
+func newCloudWatchLogClient(svc cloudwatchlogsiface.CloudWatchLogsAPI, logger *zap.Logger) *cloudWatchLogClient {
+	logClient := &cloudWatchLogClient{svc: svc,
+		logger: logger}
+	return logClient
+}
+
+// NewCloudWatchLogsClient create cloudWatchLogClient
+func NewCloudWatchLogsClient(logger *zap.Logger, awsConfig *aws.Config, sess *session.Session) LogClient {
+	client := cloudwatchlogs.New(sess, awsConfig)
+	client.Handlers.Build.PushBackNamed(handler.RequestStructuredLogHandler)
+	return newCloudWatchLogClient(client, logger)
+}
+
+//Put log events. The method mainly handles different possible error could be returned from server side, and retries them
+//if necessary.
+func (client *cloudWatchLogClient) PutLogEvents(input *cloudwatchlogs.PutLogEventsInput, retryCnt int) (*string, error) {
+	var response *cloudwatchlogs.PutLogEventsOutput
+	var err error
+	var token = input.SequenceToken
+
+	for i := 0; i <= retryCnt; i++ {
```

Comment:
Can we use callfuncwithretries? It's weird to have two ways of retrying

Reply:
Also do we even need retry? Doesn't the `queued_retry` processor work? If not, document why you are implementing retry logic by yourself

Reply:
In this switch-case, there are only 2 cases we need retry. (`InvalidSequenceTokenException` and `ResourceNotFoundException`) These 2 cases are actually handling some expected error under some circumstances (e.g. first time to publish data to a new logGroup/logStream). We use the token returned from first attempt and do an immediate retry to publish the data. We handle this inside exporter mainly because these 2 exceptions are not considered as 'real' server or client exceptions.

Reply:
> Can we use callfuncwithretries? It's weird to have two ways of retrying

This is also because there are only 2 cases which need retry. For the other cases, we just want to throw the error back to pipeline and let processor handle it. The general `callfuncwithretries` can ignore some exceptions but it does not throw the error back.